tools/ansible-pr/run

```diff
@@ -0,0 +1,230 @@
+#!/usr/bin/env ruby
+# Copyright 2017 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'optparse'
+require_relative '../../compile/core'
+require 'octokit'
+require 'time'
+
+# Class to handle Compile include to avoid Rubocop issues.
+class CompileClass
+  include Compile::Core
+end
+
+##########################################################
+# Default Options
+#
+ANSIBLE_URL = 'git@github.com:ansible/ansible.git'.freeze
+MAGICIAN_GIT = 'git@github.com:modular-magician/ansible.git'.freeze
+
+IGNORED_MODULES = %w[
+  gcp_backend_service
+  gcp_forwarding_rule
+  gcp_healthcheck
+  gcp_target_proxy
+  gcp_url_map
+].freeze
+
+ADDITIONAL_FILES = [
+  'lib/ansible/module_utils/gcp_utils.py',
+  'lib/ansible/utils/module_docs_fragments/gcp.py'
+].freeze
+
+@options = {
+  directory: "#{Dir.pwd}/build/ansible/",
+  dry_run: false,
+  # Get all PRs from one day ago.
+  date: (Time.now - (3600 * 24)).iso8601
+}
+###########################################################
+# Functions
+
+def git_run(cmd, silent = false, opts = @options)
+  silent_cmd = silent ? '>/dev/null' : ''
+  command = [
+    "pushd #{opts[:directory]} >/dev/null",
+    "#{cmd}#{silent_cmd} 2>&1",
+    "popd #{opts[:directory]} > /dev/null"
+  ]
+
+  if @options[:dry_run]
+    puts command.join('; ')
+  else
+    %x(#{command.join('; ')})
```

Comment:
Is there no better way than shelling out here?  I strongly prefer a ruby library that manages git repos for you.  Without that, I wonder if it'd be a better shell script than ruby script.

Reply:
I spent way too much time trying to work with the [rugged gem](https://github.com/libgit2/rugged).

Issues I had included:
* Getting it to work with SSH repos was incredibly hard.
* Documentation isn't great.
* It can replicate most of this script, but not all of it (finding all files not in x branch).

I'm not convinced the gem can get the job done or that it's worth the time trying to figure out if it can get the job done.

As for converting to a Bash script, how strong of an opinion do you have? My bias is that simple scripts should be Bash, complicated should be anything else. I think this script is complicated enough to warrant being in anything-but-Bash.

Reply:
If we could use gbash that would make things not so bad, but since we're already in ruby here I think it makes sense.

My concern with shelling out from a ruby script is the same as always: will this ever be run any place other than the user's own machine, e.g., on concourse? If not, I'm not too worried about an attacker naming their branch `foo; rm -rf ~ #` since they'll just be owning themselves.